sigs/jvm/rfcs/20190904-java-tensors.md

```diff
@@ -0,0 +1,301 @@
+flexible, those factories make use of heavy reflection techniques that offer poor performances and must be avoided.
+[Other factories](https://github.com/karllessard/tensorflow/blob/44ebaf12e7f196ef621413ce90c48bb9ae5f7522/tensorflow/java/src/main/java/org/tensorflow/Tensor.java#L167)
+accept a `java.nio.Buffer` for data input, which proved to be way more efficient but is not
+easy to use, especially when dealing with high rank tensors. Those factories were already planned to be replaced
+by an equivalent that accept an `NdArray` instead.
+
+Then there is a multitude of factories found in the [`Tensors` helper class](https://github.com/karllessard/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/Tensors.java),
+which make all use of the non-efficient reflective allocators discussed before, so they must be avoided in their
+actual form as well.
+
+Finally, the [`Constant` operation wrapper](https://github.com/karllessard/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/op/core/Constant.java)
+offers also a convenient way to quickly allocate a single-use tensor for creating a constant, most are still based on 
+the same non-efficient reflective allocators.
+
+For a better experience, it should be more intuitive for a user to know which factory to use and when. Basically, there
+is two main cases to cover: allocating a tensor for feeding data to the network and allocating a tensor for 
+creating a constant.
+
+#### Data Tensors
+
+Data tensors, such as those feeding a network, must be allocated explicitely by the user by providing a data type
+and the data in input. Having the factory methods directly in the tensor type classes avoid the former parameter
+while the data in input could be initialize using basic `NdArray` functionalities. For example:
+```java
+public class TInt32 extends Tensor<Integer> implements Numeric {
+
+    public static final DataType<TInt32> DTYPE = DataType.create(TInt32.class, 3, 4);
+ 
+    public static TInt32 ofShape(long... dimensionSizes) {
+        return allocate(DTYPE, Shape.make(dimensionSizes);
+    }
+}
+
+TInt32 matrix = TInt32.ofShape(2, 2).row(10, 12).row(32, 42).done();
+```
+*Note: the `row` method in the previous example does not exist in the previously presented `NdArray` interface
+and is an addendum to this RFC that can be consulted at the end of this document*
+
+Also, for convenience, nothing prevent us to have more factory methods in those classes, to allocate tensors
+of known shapes or to copy existing N-dimensional data into a new tensor.
+```java
+public class TInt32 extends Tensor<Integer> implements Numeric {
+
+   ...
+   
+   public static TInt32 scalar(Integer value) {
+       return allocate(DTYPE, Shape.make()).set(value);
+   }
+   
+   public static TInt32 vector(Integer... values) {
```

Comment:
The boxing here makes this very inefficient, and also it's not possible to pass in an `int[]`. We'd need a signature that is `int...` in addition to these methods.

Reply:
Can you please give more explanation what makes the boxed varargs not performant, other than the regular required conversion? I'm lacking a bit of knowledge in this field and I think you know a lot more about it. 

I'm just trying to avoid supporting primitive arrays in the `NdArray` API, like I used to, because it complexifies a lot the hierarchy of objects required if we want to keep control on which tensor supports which primitive array (i.e. you can pass a `int[]` to a `TInt32` but not to a `TBool`).

Reply:
The varargs construction allocates an array and writes all the values into it. So for an `f(Integer... inputs)` called as `obj.f(1000,2000,3000,4000)` it first creates an Integer object which contains the field storing 1000, then an Integer representing 2000, etc. Then it stores a reference to each of those Integers into the created Integer[4]. This means that the call allocated 5 objects on the heap. For an entry point `g(int... inputs)` called as `obj.g(1000,2000,3000,4000)` then it allocates a single int[4] and writes all the values into that array. So that's only a single object allocation. This isn't too important for short arrays like in your example, but it's really important for longer arrays.

The integers used are important too, as the JDK has a cache of the integers in the range [-127, 127] so those integers shouldn't trigger allocation. But values outside those ranges (and Floats, Doubles etc) will allocate a fresh object.

The code to convert an array of primitives into an array of unboxed primitives is a single line `IntStream.of(intArr).boxed().toArray(Integer[]::new)` and there isn't a FloatStream or BooleanStream to do the equivalent transformation for those types quickly.

I think it's important to be able to pass in a primitive array, as that's what people will be using for efficiency reasons, and if we require the intermediate boxing then it's going to generate a lot of garbage and object allocations.

Either that or we get the NDArray set(T value, int... indices) method to be extremely fast and remove all ways of creating things with arrays or varargs, so the natural way of using it is generating a blank one and then filling it with the feature extractor. If it's backed by a ByteBuffer then it's probably a quick method, but if it's writing directly into the TF Tensor then we pay the JNI cost every single time.

The split between primitives and objects is painful in current Java, as you need to use the primitives to get any kind of speed, but the lack of generics over primitives leads to ugly APIs with edges. Once specialised generics lands we should be able to get around this, but that's still a while away.

Reply:
Thanks a lot, that we very helpful. And yes, your last comment says it all:
> The split between primitives and objects is painful in current Java, as you need to use the primitives to get any kind of speed, but the lack of generics over primitives leads to ugly APIs with edges.

I guess we can start with an easy clean API principally relying on boxed types, while being careful to avoid allocation overhead like you stated, and then add the primitive arrays if performances are not reached. But like you also said: 
> This isn't too important for short arrays like in your example, but it's really important for longer arrays.
which is probably most of the cases when an `NdArray` using one of those method (`vector()`, `set()`, ...). For bulk initialize of a large set of data, probably the `NdArray.write(DataBuffer)`will be a better pick.

> If it's backed by a ByteBuffer then it's probably a quick method, but if it's writing directly into the TF Tensor then we pay the JNI cost every single time.
There should not be any difference in performances between writing to a `NdArray`or a `Tensor`as both are backed by `ByteBuffer` (currently JNI is only implied to return a reference to the `ByteBuffer`of the tensor and then everything can be done in Java space).

Reply:
When does the ByteBuffer get handed to the TF C API?

Reply:
Actually? When the tensor is allocated by the native layer, its buffer is mapped to a `ByteBuffer` (see [here](https://github.com/tensorflow/tensorflow/blob/e1667ec8563262e7fe6cd07ddc13af599e00e906/tensorflow/java/src/main/native/tensor_jni.cc#L483) and returned to the Java runtime.

So it was just per design that a user can not write directly to the tensor memory right now, something we’ll change with the new Tensor NIO interface.

BTW, @eronwright, who wrote  that code a few years ago, has join the SIG lately to offer his help on this.

Reply:
Ah, cool, I didn't realize that. Does that interfere with any of the automatic memory management it's doing under the hood (e.g. CUDA unified memory, or whatever magic happens with TPUs)?

Reply:
My blind guess is that a tensor in GPU memory is always copy to local memory before being mapped to a ByteBuffer, but Eron probably know better about it.

If that's the case, then yes there would be a performance impact caused by the copy so we should make sure to map that memory only when required (which is the case right now... I think).

Reply:
I wouldn't be surprised if a direct `ByteBuffer` could manipulate a GPU's mapped memory.   Not an expert on that.

About the whole concept of an `NDArray` that is backed by a direct byte buffer, we might find inspiration in the `V8TypedArray` in the Java bindings for V8 ([article](https://eclipsesource.com/blogs/2016/07/25/typedarrays-sharing-memory-between-v8-and-java/)).

Regarding the copying of buffer data that TF Java performs today, I believe the rationale is that TF assumes that it owns the memory.   There's some reference counting there but I haven't looked at it recently.   Meanwhile, I am very keen on introducing Netty-style reference counting to TF Java, and intend to write a proposal to that effect.